packages/hash/shared/src/ProsemirrorManager.ts

```diff
@@ -0,0 +1,530 @@
+        );
+      }),
+    );
+
+    const { tr } = currentState;
+
+    addEntityStoreAction(currentState, tr, { type: "store", payload: store });
+
+    tr.replaceWith(0, currentState.doc.content.size, newNodes);
+
+    return tr;
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async replaceNode(
+    draftBlockId: string,
+    targetComponentId: string,
+    targetVariant: BlockVariant,
+    node: ProsemirrorNode<Schema>,
+    pos: number,
+  ) {
+    const { view } = this;
+
+    // @todo consider separating the logic that uses a view to something else
+    if (!view) {
+      throw new Error("Cannot trigger replaceNodeWithRemoteBlock without view");
+    }
+
+    const [tr, newNode] = await this.createBlock(
+      targetComponentId,
+      draftBlockId,
+      targetVariant,
+    );
+
+    tr.replaceRangeWith(pos, pos + node.nodeSize, newNode);
+    view.dispatch(tr);
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async deleteNode(node: ProsemirrorNode<Schema>, pos: number) {
+    const { view } = this;
+
+    if (!view) {
+      throw new Error("Cannot trigger deleteNode without view");
+    }
+
+    const { tr } = view.state;
+
+    tr.delete(pos, pos + node.nodeSize);
+    view.dispatch(tr);
+  }
+
+  async createBlock(
+    targetComponentId: string,
+    draftBlockId: string | null,
+    targetVariant: BlockVariant,
+  ) {
+    if (!this.view) {
+      throw new Error("Cannot trigger createBlock without view");
+    }
+
+    await this.defineBlockByComponentId(targetComponentId);
+
+    const { tr } = this.view.state;
+
+    const entityProperties = targetVariant?.properties ?? {};
+    const entityStoreState = entityStorePluginState(this.view.state);
+    const blockEntity = draftBlockId
+      ? entityStoreState.store.draft[draftBlockId]
+      : null;
+
+    if (
+      (draftBlockId && !blockEntity) ||
+      (blockEntity && !isDraftBlockEntity(blockEntity))
+    ) {
+      throw new Error("draft id does not belong to a block");
+    }
+
+    let targetBlockId: string;
+
+    if (targetComponentId === blockEntity?.properties.componentId) {
+      addEntityStoreAction(this.view.state, tr, {
+        type: "updateEntityProperties",
+        payload: {
+          draftId: blockEntity.properties.entity.draftId,
+          properties: entityProperties,
+          merge: true,
+        },
+      });
+      targetBlockId = blockEntity.draftId;
+    } else {
+      const newBlockProperties =
+        blockEntity &&
+        isTextBlock(targetComponentId) &&
+        isDraftTextContainingEntityProperties(
+          blockEntity.properties.entity.properties,
+        )
+          ? /**
+             * This is supporting swapping between text blocks and persisting the
+             * existing text
+             */
+            {
+              ...entityProperties,
+              text: blockEntity.properties.entity.properties.text,
```

Comment:
Sorry, I might have missed this somewhere. I thought the structure of text block properties looked more like

```
blockEntity.properties.entity.properties.tokens
```

Are we replacing `tokens` with `text` now?

Reply:
No. `text` is the property on which legacy links are store *to* text entities *from* block data entities (which may store a heading level, for example). Blocks shouldn't link directly to text entities. There's no change here.